tests/langchain/test_langchain_model_export.py

```diff
@@ -302,6 +325,177 @@ def test_langchain_native_log_and_load_qa_with_sources_chain():
     assert model == loaded_model
 
 
+def load_retriever(persist_directory):
+    embeddings = FakeEmbeddings(size=5)
+    vectorstore = FAISS.load_local(persist_directory, embeddings)
+    return vectorstore.as_retriever()
+
+
+@pytest.mark.skipif(
+    version.parse(langchain.__version__) < version.parse("0.0.194"),
+    reason="Saving RetrievalQA chains requires langchain>=0.0.194",
+)
+def test_log_and_load_retrieval_qa_chain(tmp_path):
+    # Create the vector db, persist the db to a local fs folder
+    loader = TextLoader("tests/langchain/state_of_the_union.txt")
+    documents = loader.load()
+    text_splitter = CharacterTextSplitter(chunk_size=1000, chunk_overlap=0)
+    docs = text_splitter.split_documents(documents)
+    embeddings = FakeEmbeddings(size=5)
+    db = FAISS.from_documents(docs, embeddings)
+    persist_dir = str(tmp_path / "faiss_index")
+    db.save_local(persist_dir)
+
+    # Create the RetrievalQA chain
+    retrievalQA = RetrievalQA.from_llm(llm=OpenAI(), retriever=db.as_retriever())
+
+    # Log the RetrievalQA chain
+    with mlflow.start_run():
+        logged_model = mlflow.langchain.log_model(
+            retrievalQA,
+            "retrieval_qa_chain",
+            loader_fn=load_retriever,
+            persist_dir=persist_dir,
+        )
+
+    # Remove the persist_dir
+    shutil.rmtree(persist_dir)
+
+    # Load the chain
+    loaded_model = mlflow.langchain.load_model(logged_model.model_uri)
+    assert loaded_model == retrievalQA
+
+    # Serve the chain
+    loaded_pyfunc_model = mlflow.pyfunc.load_model(logged_model.model_uri)
+    langchain_input = {"query": "What did the president say about Ketanji Brown Jackson"}
+    with _mock_request(return_value=_mock_chat_completion_response()):
+        result = loaded_pyfunc_model.predict([langchain_input])
+        assert result == [TEST_CONTENT]
+
+    inference_payload = json.dumps({"inputs": langchain_input})
+    langchain_output_serving = {"predictions": _chat_completion_json_sample(TEST_CONTENT)}
+    with _mock_request(return_value=_MockResponse(200, langchain_output_serving)) as mock_request:
+        import mlflow.pyfunc.scoring_server as pyfunc_scoring_server
+        from mlflow.deployments import PredictionsResponse
+
+        response = pyfunc_serve_and_score_model(
+            logged_model.model_uri,
+            data=inference_payload,
+            content_type=pyfunc_scoring_server.CONTENT_TYPE_JSON,
+            extra_args=["--env-manager", "local"],
+        )
+
+        assert (
+            PredictionsResponse.from_json(response.content.decode("utf-8"))
+            == langchain_output_serving
+        )
+        assert mock_request.call_count == 3
```

Comment:
I probed what called the mock:

```
>>> mock_request.call_args_list
[call(method='get', url='http://localhost:45339/ping', params=None),
 call(method='get', url='http://localhost:45339/version', params=None),
 call(method='post', url='http://localhost:45339/invocations', data='{"inputs": {"query": "What did the president say about Ketanji Brown Jackson"}}', json=None, headers={'Content-Type': 'application/json'})]
```

Reply:
Got this error without the mock:

```
2023/07/01 12:36:06 INFO mlflow.store.db.utils: Creating initial MLflow database tables...
2023/07/01 12:36:06 INFO mlflow.store.db.utils: Updating database tables
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
2023/07/01 12:36:06 INFO mlflow.models.flavor_backend_registry: Selected backend for flavor 'python_function'
2023/07/01 12:36:06 INFO mlflow.pyfunc.backend: === Running command 'exec gunicorn --timeout=60 -b 127.0.0.1:42213 -w 1 ${GUNICORN_CMD_ARGS} -- mlflow.pyfunc.scoring_server.wsgi:app'
[2023-07-01 12:36:06 +0900] [31630] [INFO] Starting gunicorn 20.1.0
[2023-07-01 12:36:06 +0900] [31630] [INFO] Listening at: http://127.0.0.1:42213 (31630)
[2023-07-01 12:36:06 +0900] [31630] [INFO] Using worker: sync
[2023-07-01 12:36:06 +0900] [31632] [INFO] Booting worker with pid: 31632
[2023-07-01 12:36:08 +0900] [31632] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
    worker.init_process()
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/workers/base.py", line 134, in init_process
    self.load_wsgi()
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
    return self.load_wsgiapp()
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/util.py", line 359, in import_app
    mod = importlib.import_module(module)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/haru/Desktop/repositories/mlflow/mlflow/pyfunc/scoring_server/wsgi.py", line 6, in <module>
    app = scoring_server.init(load_model(os.environ[scoring_server._SERVER_MODEL_PATH]))
  File "/home/haru/Desktop/repositories/mlflow/mlflow/pyfunc/__init__.py", line 597, in load_model
    model_impl = importlib.import_module(conf[MAIN])._load_pyfunc(data_path)
  File "/home/haru/Desktop/repositories/mlflow/mlflow/langchain/__init__.py", line 681, in _load_pyfunc
    return wrapper_cls(_load_model_from_local_fs(path))
  File "/home/haru/Desktop/repositories/mlflow/mlflow/langchain/__init__.py", line 709, in _load_model_from_local_fs
    return _load_model(
  File "/home/haru/Desktop/repositories/mlflow/mlflow/langchain/__init__.py", line 610, in _load_model
    kwargs = {key: _load_from_pickle(loader_fn_path, persist_dir)}
  File "/home/haru/Desktop/repositories/mlflow/mlflow/langchain/__init__.py", line 576, in _load_from_pickle
    loader_fn = cloudpickle.load(f)
ModuleNotFoundError: No module named 'test_langchain_model_export'
```

Reply:
I manually tested the model serving, it works. https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/3408579021528539 This seems to be a test-specific error.